compiler/natives/src/crypto/sha256/sha256.go

```diff
@@ -0,0 +1,67 @@
+// +build js
+
+package sha256
+
+import (
+	"errors"
+
+	"github.com/gopherjs/gopherjs/js"
+)
+
+var ErrWebCryptoFailed = errors.New("crypto: Web Crypto API call failed")
+
+// Get the SubtleCrypto interface
+// https://www.w3.org/TR/WebCryptoAPI/
+func getSubtle() *js.Object {
+	crypto := js.Global.Get("crypto")
+	if crypto == js.Undefined {
+		crypto = js.Global.Get("msCrypto") // for IE11
+		if crypto == js.Undefined {
+			return nil
+		}
+	}
+	subtle := crypto.Get("subtle")
+	if subtle == js.Undefined {
+		subtle = crypto.Get("webkitSubtle") // for Safari
+		if subtle == js.Undefined {
+			return nil
+		}
+	}
+	return subtle
+}
+
+func jsSum256(data []byte) (res [Size]byte, err error) {
+	subtle := getSubtle()
+	if subtle == nil {
+		return res, ErrWebCryptoFailed
+	}
+
+	algo := js.M{
+		"name": "SHA-256",
+	}
+	resCh := make(chan []byte, 1)
+	errCh := make(chan error, 1)
+	promise := subtle.Call("digest", algo, data)
+	promise.Call("then", func(result *js.Object) {
+		resCh <- js.Global.Get("Uint8Array").New(result).Interface().([]byte)
+	})
+	promise.Call("catch", func(*js.Object) {
+		errCh <- ErrWebCryptoFailed
+	})
+	select {
+	case err = <-errCh:
+	case resSlice:= <-resCh:
+		copy(res[:], resSlice)
+	}
+	return
+}
+
+func Sum256(data []byte) [Size]byte {
+	res, err := jsSum256(data)
+	if err != nil {
+		// The WebCryptoAPI call failed: fallback by calling the implementation from
+		// the Go standard library
+		return _gopherjs_overriden_Sum256(data)
+	}
```

Comment:
That's a good point.
Generally speaking for crypto functions, the only cases I can think of where I would expect this to happen are:

- When the function or one of the parameters is not supported (but the function is still defined). I'm not sure which exception we will get though, I guess a NotSupportedError.
- With Chrome when the origin is insecure (NotSupportedError).
- When there is a problem with the arguments, e.g. (depending on the algorithm) decryption with a wrong key or wrong authentication tag  (According to the W3C spec, we should get an OperationError in this case).

We could try to check if the error code is NotSupportedError and then remember that the webcrypto implementation of this function shouldn't be called, but NotSupportedError may also mean that while the function itself is supported, a parameter, typically a key length or mode of operation is not supported (see https://diafygi.github.io/webcrypto-examples/), so that would be counterproductive in this case to globally avoid the JS implementation of the function.

The performance impact of calling the JS implementation when we get a NotSupportedError is about 5ms under Chrome on insecure origin: http://www.gopherjs.org/playground/#/jlVJjSdJeh